### ISSUE TITLE
Dynamoose with aws-sdw-mock?

### ISSUE THREAD

**Issue by jdcrecur** [dwyl/aws-sdk-mock#109]
We have an app running with [dynamoose](https://dynamoosejs.com/api)
Mocking the response from the database is proofing extremely difficult.

Has anyone managed to use aws-sdw-mock successfully with dynamoosejs?

```
AWS_SDK.config.region = config.awsRegion
AWS_SDK.config.accessKeyId = config.awsAccessKeyId
AWS_SDK.config.secretAccessKey = config.awsSecretAccessKey

const AWS = require('aws-sdk-mock')
AWS.setSDKInstance(AWS_SDK)
AWS.mock('DynamoDB', 'deleteItem', function (params, callback) {
  console.log('successfully deleteItem item in database')
  callback(null, 'successfully put item in database')
})
AWS.mock('DynamoDB', 'updateItem', function (params, callback) {
  console.log('successfully updateItem item in database')
  callback(null, 'successfully put item in database')
})
AWS.mock('DynamoDB', 'putItem', function (params, callback) {
  console.log('successfully putItem item in database')
  callback(null, 'successfully put item in database')
})
AWS.mock('DynamoDB', 'query', function (params, callback) {
  console.log('successfully query item in database')
  callback(null, 'successfully put item in database')
})
AWS.mock('DynamoDB', 'getItem', function (params, callback) {
  console.log('successfully getItem item in database')
  callback(null, 'successfully put item in database')
})
AWS.mock('DynamoDB', 'scan', function (params, callback) {
  console.log('successfully scan item in database')
  callback(null, 'successfully put item in database')
})
```

I then run a normal test which uses a simple query but the mock is not used and the test runs a real query to the database eg:
```
    let newStation = new Station({
      id: uuid(),
      name: name,
      position: {
        longitude: longitude,
        latitude: latitude,
      },
    })
    return await newStation.save()
```

On a simple model:
```
import dynamoose from 'dynamoose'
import config from '../../config'

export default dynamoose.model(config.modelNameStation, {
  id: {
    type: String,
    required: true,
    hashKey: true,
  },
  position: {
    latitude: {
      type: Number,
      required: true,
    },
    longitude: {
      type: Number,
      required: true,
    },
  },
  name: {
    type: String,
    required: true,
  },
},
  {
    prefix: config.dbPrefix,
    create: true,
    update: true,
  })
```

The tests are with jest eg:
```
import StationQueries from '../StationQueries'

describe('statusUpdate tests', () => {
  it('test', async () => {
    console.log('running status update')
    await StationQueries.create('asas', 4567, 4567)
    console.log('running status update done')
    expect(1).toBe(1)
  })
})
```

The only output i get from running the test is the 'running status update' && 'running status update done' and the query is actually run against the db.

**Comment by jdcrecur** [comment]
Nope. I have moved away from dynamodb though now, it is too limiting.

**Comment by Ibbus93** [comment]
Even if this is not correlated with aws-sdk-mock, I post my solution.
If you don't want to read the following explanation, you can just navigate to the [repository](https://github.com/bardarz/crud-node-serverless/), that shows a CRUD solution for an AWS Lambda application based on express.

Essentially, the project linked above is built in this way:
*  There is a file with different [routing paths](https://github.com/bardarz/crud-node-serverless/blob/master/src/routes/index.js);
*  Each route exposed refers to a function imported from a [controller](https://github.com/bardarz/crud-node-serverless/blob/master/src/controllers/item.controller.js);
*  The controller asks for data to a [Data Model](https://github.com/bardarz/crud-node-serverless/blob/master/src/models/item.model.js); 
*  Finally, the Data Model queries directly a [dynamoose schema](https://github.com/bardarz/crud-node-serverless/blob/master/src/models/schemas/item.schema.js).

The modules to mock in this case are:
*  The **req** and **res** params of the route handler in the controlller;
*  The **response of the Data Model**: in this way, there is no need to mock directly a dynamoose model because there is an upper layer who is demanded to ask the data and return them.

Finally, [this is my solution](https://github.com/bardarz/crud-node-serverless/blob/master/tests/item.controller.test.js). There is a custom mock implementation of the Get Item from the model depending on the test.

I hope this explanation will help someone dealing with mocking and testing with dynamoose.

**Comment by SethThomas** [comment]
Thanks for posting this perspective @Ibbus93.  I am going through a similar process now with dynamodb-toolbox.  If my code were calling DynamoDB directly, mocking calls to DDB makes sense.  In my case, my code doesn't call DocumentClient directly (dynamodb-toolbox does for me).  I've never worked with Dynamoose, but I suspect it does the same.

I had been giving a lot of thought to how I wanted to write my test cases.  If I mock calls to DcoumentClient, I am not only testing my code but the code of the 3rd party library that actually touches DDB.  That is _not_ my intention!  Instead, I plan on mocking out dynamodb-toolbox.  In the case of unit testing, I don't particularly care how dynamodb-toolbox uses the DocumentClient interface.  I care that `User.get` returns a User!  Thus, mocking out User.get is a more meaningful test.

Anyways, just wanted to say thanks for posting your solution.  It's nice to see a confirmation of the approach I was considering.
